src/storage/src/storage_state.rs

```diff
@@ -92,6 +96,79 @@ pub struct StorageState {
     /// equal to this frontier)
     pub sink_write_frontiers: HashMap<GlobalId, Rc<RefCell<Antichain<Timestamp>>>>,
+    /// See: [SinkHandle]
+    pub sink_handles: HashMap<GlobalId, SinkHandle>,
+}
+
+/// This maintains an additional read hold on the source data for a sink, alongside
+/// the controller's hold and the handle used to read the shard internally.
+/// This is useful because environmentd's hold might expire, and the handle we use
+/// to read advances ahead of what we've successfully committed.
+/// In theory this could be stored alongside the other sink data, but this isn't
+/// intended to be a long term solution; either this should become a "critical" handle
+/// or environmentd will learn to hold its handles across restarts and this won't be
+/// needed.
+pub struct SinkHandle {
+    downgrade_tx: UnboundedSender<Antichain<Timestamp>>,
+    _handle: JoinHandle<()>,
+}
+
+impl SinkHandle {
+    /// A new handle.
+    pub fn new(
+        persist_location: PersistLocation,
+        shard_id: ShardId,
+        persist_clients: Arc<Mutex<PersistClientCache>>,
+    ) -> SinkHandle {
+        let (downgrade_tx, mut rx) = mpsc::unbounded_channel();
+
+        let _handle = mz_ore::task::spawn(|| "Sink handle advancement", async move {
+            let client = persist_clients
+                .lock()
+                .await
+                .open(persist_location)
+                .await
+                .expect("opening persist client");
+
+            let mut read_handle: ReadHandle<SourceData, (), Timestamp, Diff> = client
+                .open_reader(shard_id)
+                .await
+                .expect("opening reader for shard");
+
+            let mut downgrade_to = read_handle.since().clone();
+            'downgrading: loop {
+                tokio::select! {
+                    downgrade = rx.recv() => {
+                        if let Some(downgrade) = downgrade {
+                            downgrade_to = downgrade;
+                        } else {
+                            // The sending end of this channel has been dropped, which means
+                            // we're no longer interested in this handle.
+                            break 'downgrading;
+                        }
+                    }
+                    _ = sleep(Duration::from_secs(1)) => {}
+                };
+                read_handle.maybe_downgrade_since(&downgrade_to).await
+            }
+
+            // Proactively drop our read hold.
+            read_handle.expire().await;
+        });
+
+        SinkHandle {
+            downgrade_tx,
+            _handle,
+        }
+    }
+
+    /// Request a downgrade of the since. This should be called regularly;
+    /// the internal task will debounce.
+    pub fn downgrade_since(&self, to: Antichain<Timestamp>) {
+        self.downgrade_tx
+            .send(to)
+            .expect("sending to downgrade task")
```

Comment:
```suggestion
        let _ = self.downgrade_tx
            .send(to);
```

(same when its a `watch` as well, this downgrade is just ignored if the task died, instead of double-panicking or whatever else)

Reply:
I think this is needed for multi-worker as well?

Reply:
I'm not sure I understand the suggestion... IIUC the only reason a send can fail is if the receiver is dropped before the sender, which I don't see any way to produce in this code without a bug?

Reply:
as i see it, the receiver will be dropped in non-active workers for multi-worker, its just luck of the draw that we end up stopping since updates before we hit this codepath, but i do think `let _` is strictly safer